pkg/controller/daemon/update.go

```diff
@@ -156,29 +158,75 @@ func (dsc *DaemonSetsController) rollingUpdate(ctx context.Context, ds *apps.Dae
+			}
 		case newPod == nil:
 			// this is a surge candidate
 			switch {
 			case !podutil.IsPodAvailable(oldPod, ds.Spec.MinReadySeconds, metav1.Time{Time: now}):
 				// the old pod isn't available, allow it to become a replacement
 				logger.V(5).Info("Pod on node is out of date and not available, allowing replacement", "daemonset", klog.KObj(ds), "pod", klog.KObj(oldPod), "node", klog.KRef("", nodeName))
+
+				node, err := dsc.nodeLister.Get(nodeName)
+				if err != nil {
+					return fmt.Errorf("couldn't get node for nodeName %q: %v", nodeName, err)
+				}
+
+				shouldRun, _ := NodeShouldRunDaemonPod(node, ds)
+				if err != nil {
+					return fmt.Errorf("couldn't check daemon pod to delete for nodeName %q: %v", nodeName, err)
+				}
+
+				if !shouldRun {
+					logger.V(5).Info("DaemonSet pod should be deleted from the node as it is not schedulable anymore", "daemonset", klog.KObj(ds), "pod", klog.KObj(oldPod), "node", klog.KRef("", nodeName))
+					oldPodsToDelete = append(oldPodsToDelete, oldPod.Name)
+					continue
+				}
+
 				// record the replacement
 				if allowedNewNodes == nil {
 					allowedNewNodes = make([]string, 0, len(nodeToDaemonPods))
 				}
 				allowedNewNodes = append(allowedNewNodes, nodeName)
 			case numSurge >= maxSurge:
+				if podutil.IsPodAvailable(oldPod, ds.Spec.MinReadySeconds, metav1.Time{Time: now}) {
+					availableNumber++
+				}
 				// no point considering any other candidates
 				continue
 			default:
 				logger.V(5).Info("DaemonSet pod on node is out of date, this is a surge candidate", "daemonset", klog.KObj(ds), "pod", klog.KObj(oldPod), "node", klog.KRef("", nodeName))
+
+				if podutil.IsPodAvailable(oldPod, ds.Spec.MinReadySeconds, metav1.Time{Time: now}) {
+					availableNumber++
+				}
+
+				node, err := dsc.nodeLister.Get(nodeName)
+				if err != nil {
+					return fmt.Errorf("couldn't get node for nodeName %q: %v", nodeName, err)
+				}
+
+				shouldRun, _ := NodeShouldRunDaemonPod(node, ds)
+				if err != nil {
+					return fmt.Errorf("couldn't check daemon pod to delete for nodeName %q: %v", nodeName, err)
+				}
+
+				if !shouldRun {
+					logger.V(5).Info("DaemonSet pod should be deleted from the node as it is not schedulable anymore", "daemonset", klog.KObj(ds), "pod", klog.KObj(oldPod), "node", klog.KRef("", nodeName))
+					shouldNotRunPodsToDelete = append(shouldNotRunPodsToDelete, oldPod.Name)
+					continue
+				}
 				// record the candidate
 				if candidateNewNodes == nil {
 					candidateNewNodes = make([]string, 0, maxSurge)
 				}
 				candidateNewNodes = append(candidateNewNodes, nodeName)
 			}
 		default:
+			if podutil.IsPodAvailable(oldPod, ds.Spec.MinReadySeconds, metav1.Time{Time: now}) || podutil.IsPodAvailable(newPod, ds.Spec.MinReadySeconds, metav1.Time{Time: now}) {
+				availableNumber++
+			}
 			// we have already surged onto this node, determine our state
 			if !podutil.IsPodAvailable(newPod, ds.Spec.MinReadySeconds, metav1.Time{Time: now}) {
```

Comment:
> Maybe it would look better if we make a isNewPodAvailable variable?

Thanks for your advice.
I think this variable can be incremented by the number of ready newPods or oldPods, so `isPodAvailable` is better than `isNewPodAvailable`.